test/sync-profile.test.js

```diff
@@ -0,0 +1,182 @@
+      sp1.joinSecureChannel('channel-sync', this.keyBase64),
+      sp2.joinSecureChannel('channel-sync', this.keyBase64)
+    ])
+
+    await Promise.all([
+      sp2.pushProfile(masq.profileDB, idCopy, publicProfile),
+      sp1.pullProfile('pass')
+    ])
+
+    expect(window.localStorage).to.have.lengthOf(2)
+    expect(await dbExists('profile-' + id)).to.be.true
+    expect(await dbExists('profile-' + idCopy)).to.be.true
+    // The two databases should be authorized
+    expect(masq.profileDB._authorized).to.have.lengthOf(2)
+    const localStorageProfile = window.localStorage.getItem('profile-' + idCopy)
+    expect(JSON.parse(localStorageProfile)).to.eql(publicProfile)
+
+    const syncedProfile = await masq2.openProfile(idCopy, 'pass')
+    expect(syncedProfile).to.eql(privateProfile)
+
+    // 2 devices should exist from now
+    expect(await masq2.getDevices()).to.have.lengthOf(2)
+
+    await waitForSync(masq.profileDB, masq2.profileDB)
+    // masq1 should also see the new device
+    const devices = await masq.getDevices()
+
+    expect(devices).to.have.lengthOf(2)
+    expect(devices[0]).to.haveOwnProperty('localKey')
+    expect(devices[1]).to.haveOwnProperty('localKey')
+    expect(devices[0].localKey).to.have.lengthOf(64)
+    expect(devices[1].localKey).to.have.lengthOf(64)
+    expect(devices[0].localKey).to.not.equal(devices[1].localKey)
+
+    let device1 = await masq.getDevice()
+    expect(device1.apps).to.have.lengthOf(1)
+    expect(device1.apps[0].id).to.exist
+    expect(device1.apps[0].key).to.have.lengthOf(64)
+    expect(device1.apps[0].localKey).to.have.lengthOf(64)
+    expect(device1.apps[0].localKey).to.equal(device1.apps[0].key)
+
+    expect(masq.appsDBs[device1.apps[0].id]).to.exist
+    expect(masq.appsDBs[device1.apps[0].id]._authorized).to.have.lengthOf(1)
+
+    // await sp2.pullApps(masq2, '-copy')
+    await new Promise((resolve) => { setTimeout(resolve, 5000) })
+
+    let device2 = await masq2.getDevice()
+    expect(device2.apps).to.have.lengthOf(1)
+    expect(device2.apps[0].id).to.exist
+    expect(device2.apps[0].key).to.have.lengthOf(64)
+    expect(device2.apps[0].localKey).to.have.lengthOf(64)
+    expect(device2.apps[0].localKey).to.not.equal(device2.apps[0].key)
+
+    expect(masq2.appsDBs[device2.apps[0].id]).to.exist
+
+    await new Promise((resolve) => { setTimeout(resolve, 5000) })
+
+    const appid = device1.apps[0].id
+    expect(masq.appsDBs[appid]._authorized).to.have.lengthOf(2)
+    expect(masq2.appsDBs[appid + '-copy']._authorized).to.have.lengthOf(2)
+
+    await masq2._createDBAndSyncApp('app2')
+
+    await new Promise((resolve) => { setTimeout(resolve, 5000) })
+
+    device1 = await masq.getDevice()
+    device2 = await masq2.getDevice()
+    expect(device1.apps).to.have.lengthOf(2)
+    expect(device2.apps).to.have.lengthOf(2)
+
+    await new Promise((resolve) => { setTimeout(resolve, 5000) })
+
+    expect(masq.appsDBs['app2' + '-copy']._authorized).to.have.lengthOf(2)
```

Comment:
this `copy` things need to be refactored

Reply:
I don't think it's that bad as it's only in the context of the tests. We absolutely need to replicate data between two different Indexeddb databases